src/ray/util/tests/pipe_logger_test.cc

```diff
@@ -71,59 +78,60 @@ TEST_P(PipeLoggerTest, PipeWrite) {
          absl::StrFormat("%d", pipe_buffer_size).data(),
          /*overwrite=*/1);
 
-  // TODO(core): We should have a better test util, which allows us to create a temporary
-  // testing directory.
+  // TODO(hjiang): We should have a better test util, which allows us to create a
+  // temporary testing directory.
   const std::string test_file_path = absl::StrFormat("%s.out", GenerateUUIDV4());
+  const std::string log_file_path1 = test_file_path;
+  const std::string log_file_path2 = absl::StrFormat("%s.1", test_file_path);
+
+  // Delete temporary file.
+  absl::Cleanup cleanup_test_file = [&log_file_path1, &log_file_path2]() {
+    EXPECT_TRUE(std::filesystem::remove(log_file_path1));
+    EXPECT_TRUE(std::filesystem::remove(log_file_path2));
+  };
 
   // Take the default option, which doesn't have rotation enabled.
-  StreamRedirectionOption logging_option{};
-  logging_option.file_path = test_file_path;
-  logging_option.rotation_max_size = 5;
-  logging_option.rotation_max_file_count = 2;
-
-  auto log_token = CreateRedirectionFileHandle(logging_option);
-  ASSERT_EQ(write(log_token.GetWriteHandle(), kLogLine1.data(), kLogLine1.length()),
-            kLogLine1.length());
-  ASSERT_EQ(write(log_token.GetWriteHandle(), kLogLine2.data(), kLogLine2.length()),
-            kLogLine2.length());
+  StreamRedirectionOption stream_redirection_opt{};
+  stream_redirection_opt.file_path = test_file_path;
+  stream_redirection_opt.rotation_max_size = 5;
+  stream_redirection_opt.rotation_max_file_count = 2;
+
+  auto stream_redirection_handle = CreateRedirectionFileHandle(stream_redirection_opt);
+  stream_redirection_handle.CompleteWrite(kLogLine1.data(), kLogLine1.length());
+  stream_redirection_handle.CompleteWrite(kLogLine2.data(), kLogLine2.length());
   // Write empty line, which is not expected to appear.
-  ASSERT_EQ(write(log_token.GetWriteHandle(), "\n", /*count=*/1), 1);
+  stream_redirection_handle.CompleteWrite("\n", /*count=*/1);
```

Comment:
if we write an empty line, log file should also contain an empty line right?

Reply:
You're right, second though it's actually more complicated:
- What if user prints multiple continuous newliners
- Loggers usually ignore the last newliner user gives, while stdout/stderr don't

It's a bug which is out of this PR's scope, left a TODO to fix; I will get it done as the highest priority